src/hashers.cpp

```diff
@@ -0,0 +1,268 @@
+		E = D; D = C; C = circularShift(30, B); B = A; A = tmp;
+	}
+
+	for (int i = 20; i < 40; ++i) {
+		const uint32_t tmp = circularShift(5, A) + (B ^ C ^ D) + E + W[i] + 0x6ED9EBA1;
+		E = D; D = C; C = circularShift(30, B); B = A; A = tmp;
+	}
+
+	for (int i = 40; i < 60; ++i) {
+		const uint32_t tmp = circularShift(5, A) + ((B & C) | (B & D) | (C & D)) + E + W[i] + 0x8F1BBCDC;
+		E = D; D = C; C = circularShift(30, B); B = A; A = tmp;
+	}
+
+	for (int i = 60; i < 80; ++i) {
+		const uint32_t tmp = circularShift(5, A) + (B ^ C ^ D) + E + W[i] + 0xCA62C1D6;
+		E = D; D = C; C = circularShift(30, B); B = A; A = tmp;
+	}
+
+	H[0] += A;
+	H[1] += B;
+	H[2] += C;
+	H[3] += D;
+	H[4] += E;
+}
+
+}
+
+std::string SHA1Hasher::encode(const std::string& input, const std::string&) const
+{
+	uint32_t H[] = {
+		0x67452301,
+		0xEFCDAB89,
+		0x98BADCFE,
+		0x10325476,
+		0xC3D2E1F0
+	};
+
+	uint8_t messageBlock[64];
+	size_t index = 0;
+
+	uint32_t length_low = 0;
+	uint32_t length_high = 0;
+	for (char ch : input) {
+		messageBlock[index++] = ch;
+
+		length_low += 8;
+		if (length_low == 0) {
+			length_high++;
+		}
+
+		if (index == 64) {
+			processSHA1MessageBlock(messageBlock, H);
+			index = 0;
+		}
+	}
+
+	messageBlock[index++] = 0x80;
+
+	if (index > 56) {
+		while (index < 64) {
+			messageBlock[index++] = 0;
+		}
+
+		processSHA1MessageBlock(messageBlock, H);
+		index = 0;
+	}
+
+	while (index < 56) {
+		messageBlock[index++] = 0;
+	}
+
+	messageBlock[56] = length_high >> 24;
+	messageBlock[57] = length_high >> 16;
+	messageBlock[58] = length_high >> 8;
+	messageBlock[59] = length_high;
+
+	messageBlock[60] = length_low >> 24;
+	messageBlock[61] = length_low >> 16;
+	messageBlock[62] = length_low >> 8;
+	messageBlock[63] = length_low;
+
+	processSHA1MessageBlock(messageBlock, H);
+
+	char hexstring[41];
+	static const char hexDigits[] = {"0123456789abcdef"};
+	for (int hashByte = 20; --hashByte >= 0;) {
+		const uint8_t byte = H[hashByte >> 2] >> (((3 - hashByte) & 3) << 3);
+		index = hashByte << 1;
+		hexstring[index] = hexDigits[byte >> 4];
+		hexstring[index + 1] = hexDigits[byte & 15];
+	}
+	return {hexstring, 40};
+}
+
+bool SHA1Hasher::verify(const std::string& input, const std::string& encoded) const
+{
+	return encode(input) == encoded;
+}
+
+const std::size_t Argon2Hasher::default_memory_cost = crypto_pwhash_argon2i_MEMLIMIT_INTERACTIVE;
+const std::size_t Argon2Hasher::default_time_cost = crypto_pwhash_argon2i_OPSLIMIT_INTERACTIVE;
+
+std::string Argon2Hasher::encode(const std::string& input, const std::string& salt) const
+{
+	return encode(input, salt, default_time_cost, default_memory_cost);
+}
+
+std::string Argon2Hasher::encode(const std::string& input, const std::string& salt, std::size_t timeCost, std::size_t memoryCost) const
+{
+	char out[crypto_pwhash_argon2i_STRBYTES];
+	if (crypto_pwhash_argon2i_str(out, input.c_str(), input.size(), timeCost, memoryCost) == 0) {
```

Comment:
I could use only `crypto_pwhash_str` here but there's no clear guarantee from libsodium that Argon2 will stay forever here, so just in case it changes.